### DISCUSSION TITLE
Build fails with VSCode for blackpill-core

### DISCUSSION THREAD

**Discussion by EngMarc** [Zer0-bit/gaggiuino#345] [General]
Please understand I am not trying to annoy you. I am building up the environment and code from scratch to make sure everything works before I go changing anything.

I was able to build everything up to the 'blackpill-core' project. It took some work to get some of the dependencies which are not noted on the project site that I saw.  One was the libusb which I was able to download and build and install.

The 'blackpill-core' project won't build with a missing max6675.h called from thermocouple.h file in gagguino/src/periphherals.  I added the include file but it still fails to build with the below error. I'm now thinking that the max6675.cpp and include file needs to be referenced for inclusion but not sure . I will defer to your expertise.

"Linking .pio/build/blackpill-core/firmware.elf
/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/blackpill-core/src/gaggiuino.ino.cpp.o: in function `loop':
gaggiuino.ino.cpp:(.text.loop+0x24): undefined reference to `MAX6675::readCelsius()'
/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: gaggiuino.ino.cpp:(.text.loop+0x698): undefined reference to `MAX6675::readCelsius()'
/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/blackpill-core/src/gaggiuino.ino.cpp.o: in function `_GLOBAL__sub_I_thermocouple':
gaggiuino.ino.cpp:(.text.startup._GLOBAL__sub_I_thermocouple+0xe): undefined reference to `MAX6675::MAX6675(signed char, signed char, signed char)'
collect2: error: ld returned 1 exit status
*** [.pio/build/blackpill-core/firmware.elf] Error 1
================================ [FAILED] Took 30.33 seconds ================================"

I did try to update the dependencies but it still doesn't pick up the include file. 
Obviously you know that the file is from:  https://github.com/adafruit/MAX6675-library.git

I wanted you to know that this project doesn't build at least from a clean install on a new system. Maybe you have something on your system that isn't in the project?

Here's the list of dependencies from VSCode:

 *  Executing task: platformio pkg list --environment blackpill-core 

Resolving blackpill-core dependencies...
Platform ststm32 @ 15.4.1 (required: ststm32 @ 15.4.1)
├── framework-arduinoststm32 @ 4.20200.221104 (required: platformio/framework-arduinoststm32 @ ~4.20200.0)
├── framework-cmsis @ 2.50700.210515 (required: platformio/framework-cmsis @ ~2.50700.0)
├── tool-dfuutil @ 1.11.0 (required: platformio/tool-dfuutil @ ~1.11.0)
├── tool-openocd @ 2.1100.211028 (required: platformio/tool-openocd @ ~2.1100.0)
├── tool-stm32duino @ 1.0.1 (required: platformio/tool-stm32duino @ ~1.0.1)
└── toolchain-gccarmnoneeabi @ 1.90201.191206 (required: platformio/toolchain-gccarmnoneeabi @ ~1.90201.0)

Libraries
├── ADS1X15 @ 0.3.7 (required: robtillaart/ADS1X15 @ 0.3.7)
├── Easy Nextion Library @ 1.0.6+sha.4bd06b2 (required: git+https://github.com/Seithan/EasyNextionLibrary.git#4bd06b2a428da3cfd17f54dfd654ff2da6867ddb)
├── FlashStorage_STM32 @ 1.2.0 (required: khoih-prog/FlashStorage_STM32 @ 1.2.0)
├── HX711 @ 0.7.5+sha.cf81533 (required: git+https://github.com/banoz/HX711.git#cf81533c54abe259ef61d457d7db6eef94f6ea25)
├── PSM @ 0.0.1+sha.8f87ae0 (required: git+https://github.com/banoz/PSM.Library.git#8f87ae0a67ace095c43bdbc661ffea72d4664b4c)
└── SimpleKalmanFilter @ 0.1.0 (required: denyssene/SimpleKalmanFilter @ 0.1.0)

**Comment by jvle** [comment]
I am seeing the failure on both "blackpill-core" and "test" with the latest code.  The build was good last week.

**Reply by EngMarc** [reply]
I think test also failed.
I looked at the platformio.ini file to see if there was an obvious item missing in the dependencies. I didn't really pick anything out right away. Seems to stem from the MAX6675 library isn't getting pulled which cascades errors causing the build to fail. 

It's not my baby so may take some time for me to find the culprit. AJ or someone more familiar with it can probably spot it right away.

**Comment by EngMarc** [comment]
I found the problem. The project definition in the platformio.ini file for 'blackpill-core'  which is missing the dependency to the MAX6675. Short term until the project is corrected you can do the following. I verified it did build with the following change.

Open the platformio.ini file and go to line 158 where you will see
lib_deps =
	robtillaart/ADS1X15@0.3.7

and change it to 

lib_deps =
	adafruit/MAX6675 library@1.1.0
	robtillaart/ADS1X15@0.3.7

**Reply by jvle** [reply]
Thanks @EngMarc.  Have you found the way to get around with "test" build errors?

**Reply by EngMarc** [reply]
Jvle, pretty sure my test compiled ok once I fixed that dependency. I did a build all and they all compiled. Try refreshing your code and update the platformio.ini file and build it. That should work if there's not anything different in the code since I last did it which was earlier today 12/4/22.

**Comment by EngMarc** [comment]
nope, I was wrong, test does not compile. There's an unresolved reference to 
PSM.h
and
ADS1X15.h

I'll look at these and see if I can fix it.
More to come and feel free to private message me or email me if you want to combine efforts and share info.

**Reply by EngMarc** [reply]
actually, I think it's a stub for testing things as it is incomplete with all the dependencies.

**Reply by jvle** [reply]
I added all lib dependencies in platformio.ini file and changed the include files to ../test/*.h.  Code compiling was good; however linking still failed

**Reply by EngMarc** [reply]
sounds like an honest effort. I'm working on figuring out how to set up test and debug environments when connected to the Arduino boards (I'm getting the STM32F411 boards). Let me know if you know or have ideas.
